### PR TITLE
feat(fast-element): add trusted types for HTML

### PR DESCRIPTION
**Pull request by EisenbergEffect**
# Description

This PR integrates the new Trusted Types web standard at all locations where user HTML could come through the system.

## Motivation & context

See #2850 for the full background.

As part of this PR a new `createHTML` method was added to the `DOM` helper. Both the template compilation process and `innerHTML` bindings now force all HTML to pass though this method. Internally, the method uses an HTML Trusted Types Policy against the HTML. The default policy is a noop. However, a custom policy that sanitizes HTML can be configured through the new `setHTMLPolicy` method on the `DOM` helper. This method can only be called once, for security reasons. Documentation has been updated with guidance around trusted type policies.

In order to implement this in such a way that the policy could be configured before templates are compiled, I had to refactor a few pieces to make template compilation lazy. This allows templates to be declared without forcing immediate compilation. The compilation now happens just-in-time for the first view instantiation. Besides addressing the timing issues related to policy configuration, this also yields a startup performance benefit in the case where there are many components or templates defined, but only a few of them are actually needed for the initial render.

## Issue type checklist

- [ ] **Chore**: A change that does not impact distributed packages.
- [ ] **Bug fix**: A change that fixes an issue, link to the issue above.
- [x] **New feature**: A change that adds functionality.
- [x] **Performance** Improves performance or memory usage.

**Is this a breaking change?**

Yes, this is technically a breaking change to the low-level APIs related to template compilation. However, this is unlikely to affect anyone at this point in time.

## Process & policy checklist

- [ ] I have added tests for my changes.
- [x] I have tested my changes.
- [x] I have updated the project documentation to reflect my changes.
- [x] I have read the [CONTRIBUTING](https://github.com/Microsoft/fast-dna/blob/master/CONTRIBUTING.md) documentation and followed the [standards](https://www.fast.design/docs/en/contributing/standards) for this project.

### CODE REVIEW

File: packages/fast-element/src/directives/directive.ts [hunk 1]
```diff
@@ -19,7 +19,7 @@ export class AttachedBehaviorDirective<T = any> extends Directive {
     }
 
     public createPlaceholder(index: number) {
-        return `${this.name}="${DOM.createInterpolationPlaceholder(index)}"`;
+        return DOM.createCustomAttributePlaceholder(this.name, index);
```
Comment (EisenbergEffect) [inline packages/fast-element/src/directives/directive.ts]: This change is unrelated to the main purpose of this PR. I just moved this code into the `DOM` helper in order to facilitate reuse and participate in the DOM's frozen characteristic.

File: packages/fast-element/src/dom.ts [hunk 2]
```diff
@@ -4,7 +4,35 @@ const markerClass = `fast-${Math.random()
...
+const fastHTMLPolicy: TrustedTypesPolicy = globalThis.trustedTypes.createPolicy(
+    "fast-html",
+    {
+        createHTML: html => html,
+    }
+);
+
+let htmlPolicy: TrustedTypesPolicy = fastHTMLPolicy;
+
+export const DOM = Object.freeze({
```
Comment (EisenbergEffect) [inline packages/fast-element/src/dom.ts]: The `DOM` object is nor frozen so that it's `createHTML` method cannot be overwritten by attackers.

File: packages/fast-element/src/template.ts [hunk 3]
```diff
@@ -14,38 +14,54 @@ export interface SyntheticViewTemplate {
     create(): SyntheticView;
 }
 
-export class HTMLTemplate extends Directive
+export class ViewTemplate extends Directive
```
Comment (EisenbergEffect) [inline packages/fast-element/src/template.ts]: Renamed for greater clarity (related to interfaces) and to disambiguate from platform template primitives.

File: packages/fast-element/src/dom.ts (hunk 2 above)
Comment (nicholasrice) [reply packages/fast-element/src/dom.ts]: nice!
